File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = -1
    n, m = len(maps), len(maps[0])
    visit = [[1]*m for _ in range(n)]
    q = deque([[0,0,1]])
    visit[0][0] = 0
    dir = [[0,1], [0,-1], [1,0], [-1,0]]
    
    while q:
        i, j, cnt = q.popleft()
        if i == n-1 and j == m-1:
            answer = cnt
            break
        for d in dir:
            ni, nj = i+d[0], j+d[1]
            if 0<=ni<n and 0<=nj<m and maps[ni][nj] == 1 and visit[ni][nj] == 1:
                visit[ni][nj] = 0
                q.append([ni,nj,cnt+1])
    return answer

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    var n = maps.length, m = maps[0].length;
    var visit = maps;
    visit[0][0] = 0;
    var q = [];
    q.push([0,0,1]);
    
    var dir = [[0,1], [0,-1], [1,0], [-1,0]];
    while (q.length > 0){
        var now = q.shift();
        var i = now[0], j = now[1], cnt = now[2];
        for (d of dir) {
            var ni = i+d[0], nj = j+d[1]
            if (0<=ni && ni<n && 0<=nj && nj<m && visit[ni][nj] === 1) {        
                if (ni === n-1 && nj === m-1) {
                    return cnt+1;
            }
                q.push([ni, nj, cnt+1]);
                visit[ni][nj] = 0
            }
        }
    }
    
    return -1;
}